core/connection-pool/src/main/java/datawave/core/common/connection/AccumuloConnectionFactoryImpl.java

```diff
@@ -0,0 +1,381 @@
+package datawave.core.common.connection;
+
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.apache.accumulo.core.client.AccumuloClient;
+import org.apache.accumulo.core.client.admin.SecurityOperations;
+import org.apache.accumulo.core.client.security.tokens.PasswordToken;
+import org.apache.accumulo.core.util.Pair;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.mutable.MutableInt;
+import org.apache.log4j.Logger;
+
+import datawave.accumulo.inmemory.InMemoryAccumuloClient;
+import datawave.core.common.cache.AccumuloTableCache;
+import datawave.core.common.result.Connection;
+import datawave.core.common.result.ConnectionPool;
+import datawave.core.common.result.ConnectionPoolProperties;
+import datawave.core.common.result.ConnectionPoolsProperties;
+import datawave.webservice.common.connection.WrappedAccumuloClient;
+
+/**
+ * An accumulo connection factory
+ */
+public class AccumuloConnectionFactoryImpl implements AccumuloConnectionFactory {
+
+    private Logger log = Logger.getLogger(this.getClass());
+
+    private final AccumuloTableCache cache;
+    private final ConnectionPoolsProperties connectionPoolsConfiguration;
+
+    private Map<String,Map<Priority,AccumuloClientPool>> pools;
+
+    private String defaultPoolName = null;
+
+    private static AccumuloConnectionFactoryImpl factory = null;
+
+    public static AccumuloConnectionFactory getInstance(AccumuloTableCache cache, ConnectionPoolsProperties config) {
+        if (factory == null) {
+            synchronized (AccumuloConnectionFactoryImpl.class) {
+                if (factory == null) {
+                    setFactory(new AccumuloConnectionFactoryImpl(cache, config));
+                }
+            }
+        }
+        return factory;
+    }
+
+    private AccumuloConnectionFactoryImpl(AccumuloTableCache cache, ConnectionPoolsProperties config) {
+        this.cache = cache;
+        this.connectionPoolsConfiguration = config;
+        log.info("Initializing AccumuloConnectionFactoryImpl with " + config.getDefaultPool() + " and " + config.getPoolNames());
+        init();
+    }
+
+    public void init() {
+        this.pools = new HashMap<>();
+
+        if (this.connectionPoolsConfiguration == null) {
+            log.error("connectionPoolsConfiguration was null - aborting init()");
```

Comment:
should this throw IllegalStateException?

Reply:
Good question.  This is identical to what is currently in use on integration though.